### ISSUE TITLE
Feature: Supporting other OSs

### ISSUE THREAD

**Issue by advdv** [timeglass/glass#11]
File monitoring is implemented differently across platforms. The current implementation uses FSEvents (OSX), let me know what other platforms you would like to see implemented

**Comment by corn13read** [comment]
Of course linux, great project!

**Comment by ryneeverett** [comment]
Seems like you'd be best off using a file system events library with adapters for all the OS's. It doesn't look like any of the Go libraries are complete, but [fsnotify](https://github.com/go-fsnotify/fsnotify) looks like it would cover just about everything other than OSX.

**Comment by advdv** [comment]
Thanks for the feedback, I'll be working on a linux implementation the coming month. Your suggestion for using fsnotify confirms my findings as well, under the hood I did a small spike already: https://github.com/timeglass/glass/blob/master/daemon/watching/monitor_fsnotify.go I just have to test it to confirm it works as expected.

**Comment by ryneeverett** [comment]
Sweet. I'd like to try it out, but that led to a couple more feedback points / feature requests:
- **Automated Tests**: Doesn't look like there are any yet. Supporting multiple OS's, VCS's etc is going to be a pain if you have to test the feature matrix manually.
- **Build from Source Instructions**: I'm assuming the release zipfile doesn't include the fsnotify stuff and I'm not a Go dev so I don't know how to install from source.

**Comment by ciryon** [comment]
Windows 10! :-)

**Comment by mihai-vlc** [comment]
+1 for windows support

**Comment by rey0bs** [comment]
+1 for Linux support, very useful !

**Comment by advdv** [comment]
Thanks for the input guys, the +1s are helpful in finding focus for the next release - polls without polls, great! :), I created a separate issues for the test&contribution improvements: https://github.com/timeglass/glass/issues/19

**Comment by advdv** [comment]
Just letting you know that the next release will focus on getting support for both windows and linux. File watching is the most obvious barrier (@see [this repo](https://github.com/timeglass/watch), others probably include permission problems

**Comment by advdv** [comment]
@Saimount @ionutvmi @corn13read @ciryon 
Last week I managed to complete both Linux and Windows support! A pre release with pre-build binaries for both platforms can be found [here](https://github.com/timeglass/glass/releases/tag/v0.5.2). I would we super grateful if you could check it out and report your experience.

**Comment by themotu** [comment]
Awesome! I'll give it a shot in linux

**Comment by mihai-vlc** [comment]
on windows the files need to have the .exe extension
why do I need to have an origin remote ?

```
Timeglass: hooks written
fatal: 'origin' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
2015/06/14 22:10:01 [Timeglass Error]: Failed to pull time data: Failed to fetch from remote 'origin' using git command [fetch origin refs/notes/time-spent:refs/notes/time-spent]: exit status 128
```

Why does the glass daemon need access from the firewall  ? (sorry didn't get the chance to look into the code)
I did notice it only shows only the elapsed minutes. The seconds are always 0.

Other then that it seems to work ok.  
The time information is appended at the end of the commit.

UPDATE: my tests were done on win 7 x64

**Comment by ciryon** [comment]
Thanks a lot for this, very appreciated! :-) I've done some limited testing on Windows 10 Pro Insider Preview (build 10130) and for me it seems to work great. All I had to do was to rename the binaries so they have .exe extensions. I'm using Cmder as my shell.

**Comment by Saimount** [comment]
Thanks a lot ! I will test your work now :)

**Comment by rey0bs** [comment]
Hi,It works well on linux. Thanks for this tool ;)
(Tested on Xubuntu 14.04 x64)

**Comment by advdv** [comment]
@ionutvmi , thanks for trying it out! When running init it will attempt to retrieve time data from the default remote: timing data is stored in a separate "notes" branch and has to be retrieved explicitly.  Additionally, whenever a timer is started it will do a version check to if you're running the latest, this is of course to speed up version rollouts. The latter probably triggered the firewall alert?

Does this seems fair? Could the experience be improved on windows?

**Comment by mihai-vlc** [comment]
I think I should be able to use it on a local git repo (without the need for a remote to be present).
It's functional but I get that nasty error message. 
You could just display a log message like: `Checking remote... not found` instead of that error message.

The version check is ok with me, the warning only shows only once anyway.

**Comment by advdv** [comment]
Ah, I understand the problem now, thanks for reporting. I created a separate issue over [here](https://github.com/timeglass/glass/issues/45). I'll close this issue as cross platform support seemed to be  working.
